bibliopixel/drivers/SPI/APA102.py

```diff
@@ -1,24 +1,43 @@
 from ... import gamma
 from .. channel_order import ChannelOrder
+from ... import data_maker
 from . base import SPIBase
 
 
 class APA102(SPIBase):
-    """Driver for APA102 based LED strips on devices like the Raspberry Pi and BeagleBone"""
+    """Driver for APA102/SK9822 based LED strips on devices like the Raspberry Pi and BeagleBone"""
 
     def __init__(self, num, gamma=gamma.APA102, **kwargs):
         super().__init__(num, gamma=gamma, **kwargs)
+        maker = kwargs.get('maker', data_maker.MAKER)
 
-        # APA102 requires latch bytes at the end
-        self._latch_bytes = [0xFF, 0, 0, 0] * ((num // 64) + 1)
+        # APA102/SK9822 requires latch bytes at the end)
+        # Many thanks to this article for combined APA102/SK9822 protocol
+        # https://cpldcpu.com/2016/12/13/sk9822-a-clone-of-the-apa102/
+        self._start_frame = 4  # start frame is [0, 0, 0, 0]
+        self._pixel_bytes = self.bufByteCount() + self.numLEDs  # 1 extra for brightness
+        self._pixel_stop = self._start_frame + self._pixel_bytes
+        self._reset_frame = 4  # for SK9822 [0, 0, 0, 0]
+        self._end_frame = ((num // 2) + 1)
+        self._packet = maker.make_packet(self._start_frame + self._pixel_bytes +
+                                         self._reset_frame + self._end_frame)
+
+        self.set_device_brightness(0xFF)  # required to setup _packet
+
+    def set_device_brightness(self, val):
+        """
+        APA102 & SK9822 support on chip brightness control allowing greater color depth.
+        APA102 superimposes a 440Hz PWM on the 19kHz base PWM to control brightness.
+        SK9822 uses a base 4.7kHz PWM but controls brightness with a variable current source.
+        Because of this SK9822 will have much less flicker at lower levels.
+        Either way, this option is better and faster than scaling in BiblioPixel
+        """
+        self._chipset_brightness = (val >> 3)  # bitshift to scale from 8 bit to 5
+        self._brightness_list = [0xE0 + self._chipset_brightness] * self.numLEDs
+        self._packet[self._start_frame + 0:self._pixel_stop:4] = self._brightness_list
 
     def _compute_packet(self):
         self._render()
-        # TODO: Do all this surgery in place instead
-        newBuf = [0xFF] * (self.bufByteCount() + self.numLEDs)
-        newBuf[1::4] = self._buf[0::3]
-        newBuf[2::4] = self._buf[1::3]
-        newBuf[3::4] = self._buf[2::3]
-        newBuf = [0, 0, 0, 0] + newBuf
-        newBuf.extend(self._latch_bytes)
-        self._packet = newBuf
+        self._packet[self._start_frame + 1:self._pixel_stop:4] = self._buf[0::3]
```

Comment:
Yeah... I was only going to fix the FILE issue and maybe add brightness. But I realized that to add brightness it was just easier to do the surgery (which I just learned you are the only one that calls it that... it's technically called slicing). But yeah, for 1000 pixels it used to take 95ms to update, and not it takes ~15ms.